### DISCUSSION TITLE
Log_logits from Model Output in GNNExplainer

### DISCUSSION THREAD

**Discussion by dloveland** [pyg-team/pytorch_geometric#2203] [Q&A]
Hello, I am a bit puzzled on part of the GNNExplainer implementation and was hoping for some guidance. As I understand it, the original paper has us taking the NLL of the class prediction after masking part of the computation adjacency matrix to assess how the prediction changed. In the pytorch_geometric code, the model outputs are named log_logits and passed directly to the loss function, but I don't see a log transform performed anywhere. To me, it looks like we are just taking the negative logits, which wouldn't make sense. Could anyone help me see what I am missing? You can find the function in https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/models/gnn_explainer.html in the explain_nodes function. Appreciate the help!

**Comment by wsad1** [comment]
Some equations for reference. pk is the probability of class k and Li is NLL loss. Note that -log(pk)=-fk-c.
![image](https://user-images.githubusercontent.com/13963626/110076696-ee3fc100-7daa-11eb-80b0-eeec3acb3fdc.png)
![image](https://user-images.githubusercontent.com/13963626/110076731-fd267380-7daa-11eb-9fcd-f8402cebbb39.png)

`log_logits = self.model(x=h, edge_index=edge_index, **kwargs)` In this line the model doesn't return a probability(pk), instead it returns fk. So your NLL loss=-fk=-log_logits=-log(pk). Essentially the model is outputting the log of the likelihood so you don't need to apply a log to that again.

**Reply by rusty1s** [reply]
Thanks for digging into this. We are indeed directly using the negative `log_logits` as loss formulation, see [here](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py#L119). This is identical to the loss formulation in `GNNExplainer`, see [here](https://github.com/RexYing/gnn-model-explainer/blob/master/explainer/explain.py#L753).

**Reply by rusty1s** [reply]
**Edit:** It might be indeed a bit weird that we are expecting the model to output log logits. Maybe we can add an additional flag to `GNNExplainer` to define this.

**Reply by dloveland** [reply]
Thank you both! I still don't see why we would expect the model to output fk (the log_logits) instead of a probability, is this a manual change we need to make within our models? I do think a flag or something would be helpful.

**Reply by wsad1** [reply]
I agree a flag might be useful. 
People either have their models return a raw score(fk), as `torch.nn.CrossEntropyLoss` expects a raw score, or probabilities(pk). But its unusual to see a model return log of likelihood (log(pk)).

**Reply by rusty1s** [reply]
This should be resolved with https://github.com/rusty1s/pytorch_geometric/pull/2209. thanks @wsad1
